File: AwsEncryptionSDK/runtimes/go/examples/cryptographicmaterialsmanager/restrictalgorithmsuite/signingsuiteonlycmm.go
package restrictalgorithmsuite

import (
	"context"
	"fmt"

	mpl "github.com/aws/aws-cryptographic-material-providers-library/releases/go/mpl/awscryptographymaterialproviderssmithygenerated"
	mpltypes "github.com/aws/aws-cryptographic-material-providers-library/releases/go/mpl/awscryptographymaterialproviderssmithygeneratedtypes"
)

type SigningSuiteOnlyCMM struct {
	approvedAlgos map[mpltypes.ESDKAlgorithmSuiteId]bool
	cmm           mpltypes.ICryptographicMaterialsManager
}

// NewSigningSuiteOnlyCMM creates a new SigningSuiteOnlyCMM
func NewSigningSuiteOnlyCMM(keyring mpltypes.IKeyring) (*SigningSuiteOnlyCMM, error) {
	// Initialize the MPL client
	materialProviders, err := mpl.NewClient(mpltypes.MaterialProvidersConfig{})
	if err != nil {
		panic(err)
	}
	// Create a DefaultCryptographicMaterialsManager
	cmmInput := mpltypes.CreateDefaultCryptographicMaterialsManagerInput{
		Keyring: keyring,
	}
	cmm, err := materialProviders.CreateDefaultCryptographicMaterialsManager(context.Background(), cmmInput)
	if err != nil {
		return nil, err
	}
	// Create list of approved algorithm
	var approvedAlgos = map[mpltypes.ESDKAlgorithmSuiteId]bool{
		mpltypes.ESDKAlgorithmSuiteIdAlgAes128GcmIv12Tag16HkdfSha256EcdsaP256: true,
		mpltypes.ESDKAlgorithmSuiteIdAlgAes192GcmIv12Tag16HkdfSha384EcdsaP384: true,
		mpltypes.ESDKAlgorithmSuiteIdAlgAes256GcmIv12Tag16HkdfSha384EcdsaP384: true,
		mpltypes.ESDKAlgorithmSuiteIdAlgAes256GcmHkdfSha512CommitKeyEcdsaP384: true,
	}
	return &SigningSuiteOnlyCMM{
		approvedAlgos: approvedAlgos,
		cmm:           cmm,
	}, nil
}

func (signingSuiteOnlyCMM *SigningSuiteOnlyCMM) GetEncryptionMaterials(input mpltypes.GetEncryptionMaterialsInput) (*mpltypes.GetEncryptionMaterialsOutput, error) {
	// Get the algorithm suite from the input
	esdkAlgorithmSuite, err := getESDKAlgorithmSuite(input.AlgorithmSuiteId)
	if err != nil {
		return nil, err
	}
	// Check if the algorithm is approved
	if !signingSuiteOnlyCMM.approvedAlgos[esdkAlgorithmSuite.Value] {
		return nil, mpltypes.AwsCryptographicMaterialProvidersException{Message: "Algorithm Suite must use Signing"}
	}
	// Delegate to the underlying CMM
	return signingSuiteOnlyCMM.cmm.GetEncryptionMaterials(input)
}

func getESDKAlgorithmSuite(algSuite mpltypes.AlgorithmSuiteId) (*mpltypes.AlgorithmSuiteIdMemberESDK, error) {
	if esdk, ok := algSuite.(*mpltypes.AlgorithmSuiteIdMemberESDK); ok {
		return esdk, nil
	}
	return nil, fmt.Errorf("algorithm suite is not ESDK type")
}

func (signingSuiteOnlyCMM *SigningSuiteOnlyCMM) DecryptMaterials(input mpltypes.DecryptMaterialsInput) (*mpltypes.DecryptMaterialsOutput, error) {
	// Get the algorithm suite from the input
	esdkAlgorithmSuite, err := getESDKAlgorithmSuite(input.AlgorithmSuiteId)
	if err != nil {
		return nil, err
	}
	// Check if the algorithm is approved
	if !signingSuiteOnlyCMM.approvedAlgos[esdkAlgorithmSuite.Value] {
		return nil, mpltypes.AwsCryptographicMaterialProvidersException{Message: "Algorithm Suite must use Signing"}
	}
	// Delegate to the underlying CMM
	return signingSuiteOnlyCMM.cmm.DecryptMaterials(input)
}

File: AwsEncryptionSDK/runtimes/rust/esdk_rust/esdk/examples/legacy/cryptographic_materials_manager/restrict_algorithm_suite/signing_suite_only_cmm.rs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use aws_esdk::mpl;
use aws_mpl_legacy::operation::decrypt_materials::DecryptMaterialsInput;
use aws_mpl_legacy::operation::decrypt_materials::DecryptMaterialsOutput;
use aws_mpl_legacy::operation::get_encryption_materials::GetEncryptionMaterialsInput;
use aws_mpl_legacy::operation::get_encryption_materials::GetEncryptionMaterialsOutput;
use aws_mpl_legacy::types::AlgorithmSuiteId;
use aws_mpl_legacy::types::EsdkAlgorithmSuiteId;
use aws_mpl_legacy::types::cryptographic_materials_manager::CryptographicMaterialsManager;
use aws_mpl_legacy::types::cryptographic_materials_manager::CryptographicMaterialsManagerRef;
use aws_mpl_legacy::types::error::Error;
use aws_mpl_legacy::types::keyring::KeyringRef;
use std::vec::Vec;

/*
 Demonstrates creating a custom Cryptographic Materials Manager (CMM).
 The SigningSuiteOnlyCMM ensures that callers use an Algorithm Suite with
 signing. This is a best practice. Read more about Digital Signing:
 https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/concepts.html#digital-sigs
 Read more about Cryptographic Materials Managers (CMMs):
 https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/concepts.html#crypt-materials-manager
*/

pub struct SigningSuiteOnlyCMM {
    approved_algos: Vec<EsdkAlgorithmSuiteId>,
    cmm: CryptographicMaterialsManagerRef,
}

impl SigningSuiteOnlyCMM {
    pub fn new(keyring: KeyringRef) -> Self {
        Self {
            approved_algos: vec![
                EsdkAlgorithmSuiteId::AlgAes128GcmIv12Tag16HkdfSha256EcdsaP256,
                EsdkAlgorithmSuiteId::AlgAes192GcmIv12Tag16HkdfSha384EcdsaP384,
                EsdkAlgorithmSuiteId::AlgAes256GcmIv12Tag16HkdfSha384EcdsaP384,
                EsdkAlgorithmSuiteId::AlgAes256GcmHkdfSha512CommitKeyEcdsaP384,
            ],
            // Create a DefaultCryptographicMaterialsManager to facilitate
            // GetEncryptionMaterials and DecryptionMaterials
            // after this CMM approves the Algorithm Suite.
            cmm: tokio::task::block_in_place(|| {
                tokio::runtime::Handle::current().block_on(async {
                    mpl()
                        .create_default_cryptographic_materials_manager()
                        .keyring(keyring)
                        .send()
                        .await
                })
            })
            .unwrap(),
        }
    }
}

impl CryptographicMaterialsManager for SigningSuiteOnlyCMM {
    fn get_encryption_materials(
        &self,
        input: GetEncryptionMaterialsInput,
    ) -> Result<GetEncryptionMaterialsOutput, Error> {
        let algorithm_suite_id: AlgorithmSuiteId = input.algorithm_suite_id.clone().unwrap();
        let esdk_algorithm_suite_id: EsdkAlgorithmSuiteId =
            if let AlgorithmSuiteId::Esdk(esdk_id) = algorithm_suite_id {
                esdk_id
            } else {
                panic!("Algorithm Suite ID is not an EsdkAlgorithmSuiteId");
            };

        if !self.approved_algos.contains(&esdk_algorithm_suite_id) {
            return Err(Error::AwsCryptographicMaterialProvidersException {
                message: "Algorithm Suite must use Signing".to_string(),
            });
        }

        tokio::task::block_in_place(|| {
            tokio::runtime::Handle::current().block_on(async {
                self.cmm
                    .get_encryption_materials()
                    .algorithm_suite_id(input.algorithm_suite_id.unwrap())
                    .commitment_policy(input.commitment_policy.unwrap())
                    .encryption_context(input.encryption_context.unwrap())
                    .max_plaintext_length(input.max_plaintext_length.unwrap())
                    .send()
                    .await
            })
        })
    }

    fn decrypt_materials(
        &self,
        input: DecryptMaterialsInput,
    ) -> Result<DecryptMaterialsOutput, Error> {
        let algorithm_suite_id: AlgorithmSuiteId = input.algorithm_suite_id.clone().unwrap();
        let esdk_algorithm_suite_id: EsdkAlgorithmSuiteId =
            if let AlgorithmSuiteId::Esdk(esdk_id) = algorithm_suite_id {
                esdk_id
            } else {
                panic!("Algorithm Suite ID is not an EsdkAlgorithmSuiteId");
            };

        if !self.approved_algos.contains(&esdk_algorithm_suite_id) {
            return Err(Error::AwsCryptographicMaterialProvidersException {
                message: "Algorithm Suite must use Signing".to_string(),
            });
        }

        tokio::task::block_in_place(|| {
            tokio::runtime::Handle::current().block_on(async {
                self.cmm
                    .decrypt_materials()
                    .algorithm_suite_id(input.algorithm_suite_id.unwrap())
                    .commitment_policy(input.commitment_policy.unwrap())
                    .encryption_context(input.encryption_context.unwrap())
                    .encrypted_data_keys(input.encrypted_data_keys.unwrap())
                    .send()
                    .await
            })
        })
    }
}
